### ISSUE TITLE
I want to tail the output of a command

### ISSUE THREAD

**Issue by kablamo** [RexOps/Rex#644]
And I would like to get stdout and stderr merged instead of separate

**Comment by ferki** [comment]
Hi @kablamo, for the tailing part there is Rex::Commands::Tail, you can find an example of using it here: http://www.rexify.org/howtos/tail.html

Merging STDERR and STDOUT is also possible AFAIK, but I need to look it up how can it be done exactly (I'm on the phone now^^).

**Comment by kablamo** [comment]
I guess I can add 2>&1 to the end of the command.  thats a pretty simple solution :)

**Comment by ferki** [comment]
Yeah, that could work if you don't need strict separation of STDOUT and STDERR. On the other hand, I was thinking of more like somehting as specifying a callback for `tail` (as if it were a simple `run` command):
```perl
task 'syslog', sub {
   tail '/var/log/syslog', sub {
    my ($stdout, $stderr) = @_;
    print "STDOUT> $stdout\n";
    print "STDERR> $stderr\n";
   };
 };
```

I haven't tested the above, but judging by a quick glance at the code, it should work™ :)

**Comment by krimdomu** [comment]
@kablamo as mentioned in irc, i've written a small module that can stream the output of a command:

http://modules.rexify.org/module/Rex::Commands::StreamRun

You need `IPC::Open3::Simple` module for this. `rexify --use Rex::Commands::StreamRun` should install this via `cpanm` if cpanm is available and the module is not installed already.

**Comment by krimdomu** [comment]
Because of the default connection mode of Net::OpenSSH, i think in one of the next releases we can refactor the internal run() command so that it can do this as well.
With the old connection mode it was pretty complicated to do this, this is why this is not working at the moment.

**Comment by krimdomu** [comment]
The code can be found here: https://github.com/RexOps/rex-recipes/tree/1.1/Rex/Commands/StreamRun

**Comment by kablamo** [comment]
Ok I used `stream_run()`.  I also added some logic to use `stream_run()` anytime the task is run against a single server.  Any other time it uses `run()` and does not stream output.

**Comment by kablamo** [comment]
Also `stream_run()` doesn't do some stuff like source the `.profile`

**Comment by ferki** [comment]
@kablamo: thanks for the patience on this issue!

## TL;DR

The underlying issue about draining STDOUT and STDERR of executed commands has been fixed in 1.4.0, but this issue wasn't closed back then. Closing it now :)

Anyone interested in some of the related ideas mentioned in this issue, please reach out and start new discussions.

## Longer version

The `continuous_read` option for the `run` command has been originally added in version 0.43.2. That implementation did not support draining STDOUT and STDERR of the executed processes in parallel, and it could lead to issues like #756 when we started to add multiple possible connection methods (e.g. Net::OpenSSH).

Various workaround were available, like calling `Rex::Config->set_no_tty(1);` or to use Rex::Commands::StreamRun from the rex-recipes repo.

Eventually, the underlying limitation has been fixed through a series of patches, and landed in the 1.3.3_04 development release of Rex, and thus also in the standard 1.4.0 release shortly afterwards.

It is possible to tail the output of a running command line-by-line like this with core Rex:

```perl
run 'mycommand', continuous_read => sub {
    my $line = shift;
    say $line;
};
```

This issue was not closed at the time, so I'm closing it now.

ps.: There has been considerable amount of changes around the project since the opening of this issue, and there are also a few slightly related ideas mentioned above. I would like to kindly ask any readers with interest in them to reach out on one of the [support channels](https://www.rexify.org/support/index.html) and/or open a [discussion](https://github.com/RexOps/Rex/discussions/new?category=ideas) about it.
